### DISCUSSION TITLE
Trained ChatGPT to use SeleniumVBA :)

### DISCUSSION THREAD

**Discussion by 6DiegoDiego9** [GCuser99/SeleniumVBA#57] [Ideas]
In the past weeks I used ChatGPT and was surprised to see how it can guess even SeleniumVBA code with zero training on it.

This night I had an idea: how far can ChatGPT go if we give it some training instructions about it?

So here is my first try:

![image](https://user-images.githubusercontent.com/109476941/211398636-a8bd6b40-4b5c-405b-8135-40adec8d8601.png)

I wonder if we should start thinking to create a function called like "ChatGPTpromptToClipboard" that creates this training preset text for us on the fly  :-)

**Comment by GCuser99** [comment]
@6DiegoDiego9, so you gave ChatGPT the descriptions ("D") above, and it created the two code sequences without any supervised training on your part? I've been hearing about this and have experience with AI in my distant past, but quite honestly, this is very impressive!

So what would your ChatGPTpromptToClipboard do exactly? What would be the inputs and outputs? Would it require a reference to third party software? I'm interested as long as it does not place any additional burden on a user that does not plan to use this feature...

Thanks for sharing!

**Comment by 6DiegoDiego9** [comment]
Yes, it's mind blowing! A big step ahead compared to other public AI bots and even to GitHub Copilot (also made by OpenAI, in collab. with GitHub) which also can generate VBA code, although it's somewhat limited to autocompleting existing code while you write it.

ChatGPT (by OpenAI) is an evolution to InstructGPT, which is trained to follow instructions in natural (or whatever, even code) language in a prompt and provide a detailed response, and can also answer follow-up questions.

Yes, you can simply start a fresh new chat with it at https://chat.openai.com, write an initial prompt like my first with the "D" (Diego) and it will give long and detailed answers like you saw, in English (or whatever local language) text and VBA code. No other actions/training needed!

Actually you could also try to ask the code straight away without any preamble instructions about what SeleniumVBA is but, since ChatGPT is trained on Internet data up to 2021, you'll get an answer like this:

![image](https://user-images.githubusercontent.com/109476941/211639360-37c8affc-053b-4ff9-a8fd-54cc5f4fc453.png)

**Comment by GCuser99** [comment]
@6DiegoDiego9, how much training pretext can you give ChatGPT? From your last example above, I wonder if you will even need the pre-text once they update the training set time window to include our SeleniumVBA development. This is very exciting - thanks for looking into and please update of status!

**Comment by 6DiegoDiego9** [comment]
I made a draft script generating one line for each method and property of SeleniumVBA, plenty of lines, without the arguments, and ChatGPT answered with the "too long" error.

I was about to ask ChatGPT to generate a Chrome/Edge Extension for me featuring a character counter for the prompt input box, to discover its limits, and while I have seen that this is feasible thought, I then decided to suspend that because:

- after seeing just about 10-20 extensions out of the 100+ already available, I already found 1 or 2 extensions that add the counter, although it's only for the ChatGPT output

- I found a couple of projects, although still in alpha state, that could provide a way to overcome the prompt size limit: LangChain and GPT Index

- and yes, the issue could no longer exist once they update the training set time window including SeleniumVBA GitHub pages

**Comment by GCuser99** [comment]
@6DiegoDiego9, a similar experience to yours... https://www.journalofaccountancy.com/news/2023/jan/using-chatgpt-with-excel.html

I can't stop saying it - mind blowing!

**Comment by GCuser99** [comment]
Just for the fun of it, I used chatGPT to help generate the wiki topics of [Explicit and Implicit Waits](https://github.com/GCuser99/SeleniumVBA/wiki#explicit-and-implicit-waits) and [Executing JavaScript Code](https://github.com/GCuser99/SeleniumVBA/wiki#executing-javascript-code). I'm sure I could do better with practice on how to ask/frame the questions. And I had to do some editing. But it seems like a great tool to get started! :-)

**Comment by 6DiegoDiego9** [comment]
Woohoo!
A GitHub developer used LangChain in a webpage that does the job of adding a personal document up to 200MB(!) in size to the knowledge of GPT3:
https://knowledgegpt.streamlit.app/

I tested it by giving to it the wiki page of SeleniumVBA:

<img width="519" alt="image" src="https://user-images.githubusercontent.com/109476941/217182251-3d239cba-1162-4446-af74-50728fa2b082.png">




The other good news of the week is that, according to some leaked screenshots ([example](https://pbs.twimg.com/media/FoBPUMAagAIibay?format=png&name=900x900)) of the new Bing, its dataset seems to be updated to last months, unless they use the Bing search engine to (sub optimally) integrate the recent data, like Perplexity.ai (with Bing) and ChatSonic (with Google) do.

**Comment by 6DiegoDiego9** [comment]
Monday I had access to the new Bing, integrating an enhanced ChatGPT.

Although it still seems to be trained with the same data of the standard ChatGPT, that is limited to 2021, and it still makes lots of errors and takes many seconds to read Bing results on each query, the out-of-the-box results can be interesting:

![image](https://user-images.githubusercontent.com/109476941/219476619-af8f48ac-5b49-4021-9d01-d3644d574bd8.png)

**Reply by GCuser99** [reply]
I like that it gives references - very cool! It is definitely combining the pre-2021 training with post-2021 references... Can you ask it to finish the Wiki for me? :-)
